File: 回溯/131.分割回文串.go
func partition(s string) [][]string {
	res := make([][]string, 0)
	path := make([]string, 0)
	var dfs func(start int)
	dfs = func(start int) {
		if start == len(s) {
			res = append(res, slices.Clone(path))
			return
		}
		for i := start; i < len(s); i++ {
			if isPalindrome(s, start, i) {
				path = append(path, string(s[start:i+1]))
				dfs(i + 1)
				path = path[:len(path)-1]
			}
		}
	}
	dfs(0)
	return res
}

func isPalindrome(s string, left, right int) bool {
	for left < right {
		if s[left] != s[right] {
			return false
		}
		left++
		right--
	}
	return true
}

File: 回溯/131.分割回文串.java
class Solution {
    List<List<String>> res;
    List<String>path;
    public List<List<String>> partition(String s) {
        res = new ArrayList<>();
        path = new ArrayList<>();
        dfs(s,0);
        return res;
    }
    void dfs(String s,int start){
        if(start == s.length()){
            res.add(new ArrayList<>(path));
            return;
        }
        for(int i=start;i<s.length();i++){
            if(isPalind(s,start,i)){
                String str = s.substring(start,i+1);
                path.add(str);
                dfs(s,i+1);
                path.removeLast();
            }
        }
    }

    boolean isPalind(String s,int left,int right){
        while(left<right){
            if(s.charAt(left++)!=s.charAt(right--)){
                return false;
            }
        }
        return true;
    }
}